core/org.openjdk.jmc.common/src/main/java/org/openjdk/jmc/common/IMCFrame.java

```diff
@@ -49,45 +55,110 @@ public interface IMCFrame {
 	/**
 	 * Frame compilation types.
 	 */
-	public enum Type {
-	/**
-	 * The frame was executed as native code compiled by the Java JIT compiler.
-	 */
-	JIT_COMPILED,
-	/**
-	 * The frame was executed as interpreted Java byte code.
-	 */
-	INTERPRETED,
-	/**
-	 * The frame was executed as code that was inlined by the Java JIT compiler.
-	 */
-	INLINED,
-	/**
-	 * The frame was executed as native code, most probably a C function.
-	 */
-	NATIVE,
-	/**
-	 * The frame was executed as native code compiled from C++.
-	 */
-	CPP,
-	/**
-	 * The frame was executed as kernel native code.
-	 */
-	KERNEL,
-	/**
-	 * The frame compilation type is unknown.
-	 */
-	UNKNOWN;
+	final class Type {
+		/**
+		 * The frame was executed as native code compiled by the Java JIT compiler.
+		 */
+		public static final Type JIT_COMPILED = new Type("JIT_COMPILED"); //$NON-NLS-1$
+
+		/**
+		 * The frame was executed as interpreted Java byte code.
+		 */
+		public static final Type INTERPRETED = new Type("INTERPRETED"); //$NON-NLS-1$
+
+		/**
+		 * The frame was executed as code that was inlined by the Java JIT compiler.
+		 */
+		public static final Type INLINED = new Type("INLINED"); //$NON-NLS-1$
 
-		private String name;
+		/**
+		 * The frame was executed as native code, most probably a C function
+		 */
+		public static final Type NATIVE = new Type("NATIVE"); //$NON-NLS-1$
 
-		private Type() {
-			name = Messages.getString("IMCFrame_Type_" + toString()); //$NON-NLS-1$
+		/**
+		 * The frame was executed as native code compiled from C++
+		 */
+		public static final Type CPP = new Type("CPP"); //$NON-NLS-1$
+
+		/**
+		 * The frame was executed as kernel native code
+		 */
+		public static final Type KERNEL = new Type("KERNEL"); //$NON-NLS-1$
+
+		/**
+		 * The frame compilation type is unknown.
+		 */
+		public static final Type UNKNOWN = new Type("UNKNONW"); //$NON-NLS-1$
+
+		private static final String MSG_PREFIX = "IMCFrame_Type_";
+
+		/*
+		 * Maximum number of items the TYPE_CACHE will hold. The assumption is that the number of
+		 * 'dynamic' frame types will be small, typically in ones rather than tens and as such 100
+		 * items in the type cache should cover the vast majority of use cases.
+		 */
+		private static final int TYPE_CACHE_MAX_SIZE = 100;
+		/*
+		 * A helper cache for the unrecognized frame types to reduce the amount of allocated
+		 * instances. The expectation is that the number of unrecognized frame types will be very
+		 * small, usually zero, so the memory overhead of the cache stays negligible.
+		 */
+		private static final Map<String, Type> TYPE_CACHE = new LinkedHashMap<String, Type>() {
+			@Override
+			protected boolean removeEldestEntry(Map.Entry eldest) {
```

Comment:
Should be: `protected boolean removeEldestEntry(Map.Entry<String, Type> eldest) {`